torchrl/modules/models/rlhf.py

```diff
@@ -0,0 +1,97 @@
+from pathlib import Path
+
+import torch
+from torch import nn as nn
+from transformers import GPT2LMHeadModel, GPT2TokenizerFast
+
+
+class GPT2RewardModel(nn.Module):
+    """Wrapper around GPT2-like models to enable their use as reward models.
+
+    This wrapper replaces the language modelling head of the GPT2 model with a new
+    linear layer with 1 output that can be used as a reward signal. It also exposes the
+    method ``compute_reward_loss`` which calculates the reward loss by comparing two
+    batches, one chosen, one rejected.
+    """
+
+    def __init__(self, model_path=None):
+        super().__init__()
+        if model_path:
+            model = GPT2LMHeadModel.from_pretrained(model_path, return_dict=False)
+        else:
+            model = GPT2LMHeadModel(GPT2LMHeadModel.config_class())
+
+        self.config = model.config
+        self.transformer = model.transformer
+
+        # replace last layer with the reward layer
+        self.lm_head = nn.Linear(self.config.n_embd, 1, bias=False)
+        self.PAD_ID = GPT2TokenizerFast.from_pretrained("gpt2").eos_token_id
+
+    def forward(self, input_ids=None, attention_mask=None):
+        """ Returns a tuple (rewards, end_scores) where `rewards` contains all rewards computed at each timestep, `end_scores` contains the reward computed at the last-non-padding token
+        """
+        outputs = self.transformer(input_ids=input_ids, attention_mask=attention_mask)
+        hidden_states = outputs[0]
+        rewards = self.lm_head(hidden_states).squeeze(-1)
+        end_scores = []
+        bs = input_ids.shape[0]
+
+        for i in range(bs):
+            pad_inds = (input_ids[i] == self.PAD_ID).nonzero()
+            first_pad_ind = (
+                pad_inds[0].item() if len(pad_inds) > 0 else input_ids.shape[1]
+            )
+            end_scores.append(rewards[i, first_pad_ind - 1])
+
+        return rewards, torch.stack(end_scores)
+
+    @staticmethod
+    def compute_reward_loss(chosen_batch, rejected_batch, pad_token_id=50256):
+        """Compute the reward loss given a chosen and rejected batch.
+
+        The loss is computed as loss = -log_sigmoid(chosen_reward - rejected_reward).
+        This loss is small when the reward model favours the chosen data and large if
+        the model favours the rejected data.
+        Note: the loss is computed excluding the common "prefix" subsequence to effectively disregard contribution of the original prompt.
+        """
+        chosen_ids = chosen_batch.input_ids
+        rejected_ids = rejected_batch.input_ids
+        chosen_rewards = chosen_batch.rewards
+        rejected_rewards = rejected_batch.rewards
+
+        bs = chosen_rewards.shape[0]
+        loss = 0
+
+        for i in range(bs):
```

Comment:
I'd be surprised that there is no more optimal solution than a loop with a bunch of calls to nonzero.
We don't need to optimize it now but let's make sure that we mark it as improvable